### PR TITLE
server: fix cannot update vm if it uses unconstraint dynamic offering

### LINKED ISSUE #6780: Failed to update VM, due to: Invalid cpu speed value, specify a value between 1 and 2147483647 - UI bug
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
UI
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0, 4.17.1
~~~

##### CONFIGURATION

~~~
NA
~~~
##### OS / ENVIRONMENT

Ubuntu 20.04, CentOS 7.9

##### SUMMARY
~~~
If I try to edit any settings under instance, I always get the same error, no matter if is not related to CPU.
~~~
##### STEPS TO REPRODUCE

~~~
There are no specific steps, just try to edit or add new value.
~~~

##### EXPECTED RESULTS

~~~
No error
~~~

##### ACTUAL RESULTS

~~~
2022-09-28 11:31:07,972 ERROR [c.c.a.ApiServer] (qtp1462044018-8587:ctx-4c809300 ctx-86f28f31) (logid:09b4a52b) unhandled exception executing api command: [Ljava.lang.String;@768c17d5
com.cloud.utils.exception.CloudRuntimeException: Failed to update VM, due to: Invalid cpu speed value, specify a value between 1 and 2147483647
        at org.apache.cloudstack.api.command.user.vm.UpdateVMCmd.execute(UpdateVMCmd.java:271)
        at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:163)
        at com.cloud.api.ApiServer.queueCommand(ApiServer.java:776)
        at com.cloud.api.ApiServer.handleRequest(ApiServer.java:600)
        at com.cloud.api.ApiServlet.processRequestInContext(ApiServlet.java:327)
        at com.cloud.api.ApiServlet$1.run(ApiServlet.java:145)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:52)
        at com.cloud.api.ApiServlet.processRequest(ApiServlet.java:142)
        at com.cloud.api.ApiServlet.doGet(ApiServlet.java:96)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
        at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1450)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:763)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:400)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:645)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:392)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.cloud.exception.InvalidParameterValueException: Invalid cpu speed value, specify a value between 1 and 2147483647
        at com.cloud.vm.UserVmManagerImpl.validateCustomParameters(UserVmManagerImpl.java:1179)
        at com.cloud.vm.UserVmManagerImpl.verifyVmLimits(UserVmManagerImpl.java:2661)
        at com.cloud.vm.UserVmManagerImpl.updateVirtualMachine(UserVmManagerImpl.java:2767)
        ... 58 more
~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #6780

The issue can be reproduced by
(1) create a unconstraint dynamic offering
(2) deploy a vm and stop it
(3) update vm settings

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java [hunk 1]
```diff
@@ -327,7 +327,7 @@ public void setDynamicFlag(boolean isdynamic) {
     }
 
     public boolean isCustomCpuSpeedSupported() {
-        return isCustomized() && getDetail("minCPU") != null;
+        return isCustomized() && getDetail("minCPU") == null;
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java:330]: it seems strange that the minimun cpu speed *must* be null and also that there are no conditions on the maximum cpu speed. Do you why this works @weizhouapache ?

[the remark this answers is not in the document]
Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java]: good, thanks @shwstppr 
committed

**Review by shwstppr — APPROVED** [review]

File: engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java (hunk 1 above)
Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java]: @DaanHoogland 
as suggested by @shwstppr , use `speed == null` instead, now it looks more clear

**Comment by weizhouapache** [comment]
@blueorangutan package

**Review by stephankruggg — APPROVED** [review]

File: engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java (hunk 1 above)
Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/service/ServiceOfferingVO.java]: @DaanHoogland 
there are two types of dynamic service offering
(1) constraint: speed is fixed. there are max/min cpu and memory.
(2) uncontraint: cpu/speed/memory are all dynamic.

the `isCustomCpuSpeedSupported` should be true, if the offering is unconstraint (speed is not set).
`getDetail("minCPU") != null` is a way to determine if the offering is uncontraint. `speed == null` is a better and simple way.

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
lgtm, tested, bug not reproduced

**Comment by meisammaleki** [comment]
Hello
 I got the same error message when try to change the instance setting, Is this issue resolved on v4.18?

**Comment by weizhouapache** [comment]
> Hello
>  I got the same error message when try to change the instance setting, Is this issue resolved on v4.18?

Yes, try with 4.18.0.0


> our version is 4.17.2

**Pull request merged**